File: framework/src/main/java/com/software/dm/swallow/stormy/security/match/bdia/common/ApplicationVagueContainer.java
package com.software.dm.swallow.stormy.security.match.bdia.common;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

import com.esotericsoftware.kryo.Kryo;
import com.software.dm.swallow.stormy.algoac.VagueAnalysisFactroy;
import com.software.dm.swallow.stormy.algoac.impl.VagueStateCharacter;
import com.software.dm.swallow.stormy.algoac.inter.AbstractVagueState;
import com.software.dm.swallow.stormy.algoac.inter.VagueState;
import com.software.dm.swallow.stormy.algoac.pojo.VagueResultEntity;
import com.software.dm.swallow.stormy.security.match.bdia.bean.T_basic_type_rel;
import com.software.dm.swallow.stormy.security.match.bdia.bean.T_extract_rule;
import com.software.dm.swallow.stormy.security.match.bdia.bean.T_ip_rule;
import com.software.dm.swallow.stormy.security.match.bdia.bean.T_theme_basic_type_rel;
import com.software.dm.swallow.stormy.security.match.bdia.bean.T_theme_url_rule;
import com.software.dm.swallow.stormy.security.match.bdia.bean.T_url_rule;

/**
 * @author DM
 * @version v1.0.0.1
 * @Description
 * @date 2017
 */
public final class ApplicationVagueContainer {

    private final Map<String, List<Object>> domainEqualsMap = new HashMap<String, List<Object>>();
    private final VagueAnalysisFactroy afDomain = new VagueAnalysisFactroy();
    private final VagueAnalysisFactroy afRule = new VagueAnalysisFactroy();
    private final Map<T_extract_rule, Pattern> extractPatternRuleMap = new HashMap<T_extract_rule, Pattern>();
    private final Map<T_extract_rule, Pattern> extractDoMainPatternRuleMap = new HashMap<T_extract_rule, Pattern>();
    private final Map<String, T_ip_rule> iPMap = new HashMap<String, T_ip_rule>();

    private final Map<Long, List<T_basic_type_rel>> basicTypeRelMap = new HashMap<Long, List<T_basic_type_rel>>();
    private final Map<Long, List<T_theme_basic_type_rel>> themeBasicTypeRelMap = new HashMap<Long, List<T_theme_basic_type_rel>>();

    private final List<T_url_rule> urlRuleList = new ArrayList<T_url_rule>();
    private final List<T_theme_url_rule> themeUrlRuleList = new ArrayList<T_theme_url_rule>();
    private final List<T_extract_rule> extractRuleList = new ArrayList<T_extract_rule>();

    public ApplicationVagueContainer() {
        // TODO Auto-generated constructor stub
    }

    public void clear() {
        // this.urlRuleList.clear();
        // this.themeUrlRuleList.clear();
        // this.extractRuleList.clear();
        this.basicTypeRelMap.clear();
        this.themeBasicTypeRelMap.clear();
    }

    public static Kryo getKryo() {
        Kryo kryo = new Kryo();
        kryo.setRegistrationRequired(true);

        kryo.register(ApplicationVagueContainer.class);
        kryo.register(VagueAnalysisFactroy.class);
        kryo.register(VagueStateCharacter.class);
        kryo.register(HashMap.class);
        kryo.register(ArrayList.class);
        kryo.register(HashSet.class);

        kryo.register(VagueResultEntity.class);
        kryo.register(T_extract_rule.class);
        kryo.register(T_basic_type_rel.class);
        kryo.register(T_theme_url_rule.class);
        kryo.register(T_theme_basic_type_rel.class);
        kryo.register(T_basic_type_rel.class);
        kryo.register(T_url_rule.class);
        kryo.register(T_ip_rule.class);
        kryo.register(AbstractVagueState.class);
        kryo.register(VagueState.class);

        kryo.register(String[].class);
        kryo.register(Integer[].class);

        kryo.register(Pattern.class);
        kryo.register(Pattern[].class);
        return kryo;
    }

    public Map<String, List<Object>> getDomainEqualsMap() {
        return domainEqualsMap;
    }

    public VagueAnalysisFactroy getAfDomain() {
        return afDomain;
    }

    public VagueAnalysisFactroy getAfRule() {
        return afRule;
    }

    // public Map<String, Set<T_extract_rule>> getExtractStringRuleMap() {
    // return extractStringRuleMap;
    // }

    public Map<T_extract_rule, Pattern> getExtractPatternRuleMap() {
        return extractPatternRuleMap;
    }

    public Map<T_extract_rule, Pattern> getExtractDoMainPatternRuleMap() {
        return extractDoMainPatternRuleMap;
    }

    public Map<String, T_ip_rule> getiPMap() {
        return iPMap;
    }

    public Map<Long, List<T_basic_type_rel>> getBasicTypeRelMap() {
        return basicTypeRelMap;
    }

    public Map<Long, List<T_theme_basic_type_rel>> getThemeBasicTypeRelMap() {
        return themeBasicTypeRelMap;
    }

    public List<T_url_rule> getUrlRuleList() {
        return urlRuleList;
    }

    public List<T_theme_url_rule> getThemeUrlRuleList() {
        return themeUrlRuleList;
    }

    public List<T_extract_rule> getExtractRuleList() {
        return extractRuleList;
    }

}

File: framework/src/main/scala/com/software/dm/swallow/stormy/scala/security/addrlib/bdia/common/ApplicationVagueContainer.scala
package com.software.dm.swallow.stormy.scala
package security.addrlib
package bdia
package common

//import java.util
//import java.{ListBuffer, HashMap, List, Map}
//import java.regex.Pattern

import java.util.regex.Pattern

import com.esotericsoftware.kryo.Kryo
import algoac.VagueAnalysisFactroy
import algoac.impl.VagueStateCharacter
import algoac.pojo.VagueResultEntity
import bean.{T_url_rule, _}
import algoac.inter.{AbstractVagueState, VagueState}

import scala.collection.mutable._

object ApplicationVagueContainer {

  def getKryo: Kryo = {
    val kryo = new Kryo
    kryo.setRegistrationRequired(true)
    kryo.register(classOf[ApplicationVagueContainer])
    kryo.register(classOf[VagueAnalysisFactroy])
    kryo.register(classOf[VagueStateCharacter])
    kryo.register(classOf[HashMap[_, _]])
    kryo.register(classOf[ListBuffer[_]])
    kryo.register(classOf[HashSet[_]])
    kryo.register(classOf[VagueResultEntity])
    kryo.register(classOf[T_extract_rule])
    kryo.register(classOf[T_basic_type_rel])
    kryo.register(classOf[T_theme_url_rule])
    kryo.register(classOf[T_theme_basic_type_rel])
    kryo.register(classOf[T_basic_type_rel])
    kryo.register(classOf[T_url_rule])
    kryo.register(classOf[T_ip_rule])
    kryo.register(classOf[AbstractVagueState[_]])
    kryo.register(classOf[VagueState[_]])
    kryo.register(classOf[Array[String]])
    kryo.register(classOf[Array[Integer]])
    kryo.register(classOf[Pattern])
    kryo.register(classOf[Array[Pattern]])


    kryo.register(classOf[scala.collection.immutable.$colon$colon[_]])
    kryo.register(scala.collection.immutable.Nil.getClass)
    kryo
  }
}

class ApplicationVagueContainer {

  private val domainEqualsMap: Map[String, ListBuffer[Any]] = new HashMap[String, ListBuffer[Any]]
  private val afDomain: VagueAnalysisFactroy = new VagueAnalysisFactroy
  private val afRule: VagueAnalysisFactroy = new VagueAnalysisFactroy
  private val extractPatternRuleMap: HashMap[T_extract_rule, Pattern] = new HashMap[T_extract_rule, Pattern]
  private val extractDoMainPatternRuleMap: HashMap[T_extract_rule, Pattern] = new HashMap[T_extract_rule, Pattern]
  private val iPMap: HashMap[String, T_ip_rule] = new HashMap[String, T_ip_rule]

  private val basicTypeRelMap: HashMap[Long, ListBuffer[T_basic_type_rel]] = new HashMap[Long, ListBuffer[T_basic_type_rel]]
  private val themeBasicTypeRelMap = new HashMap[Long, ListBuffer[T_theme_basic_type_rel]]

  private val urlRuleList: ListBuffer[T_url_rule] = new ListBuffer[T_url_rule]
  private val themeUrlRuleList: ListBuffer[T_theme_url_rule] = new ListBuffer[T_theme_url_rule]
  private val extractRuleList: ListBuffer[T_extract_rule] = new ListBuffer[T_extract_rule]


  def clear(): Unit = { // this.urlRuleList.clear();
    // this.themeUrlRuleList.clear();
    // this.extractRuleList.clear();
    this.basicTypeRelMap.clear()
    this.themeBasicTypeRelMap.clear()
  }


  def getDomainEqualsMap: Map[String, ListBuffer[Any]] = domainEqualsMap

  def getAfDomain: VagueAnalysisFactroy = afDomain

  def getAfRule: VagueAnalysisFactroy = afRule

  // public Map<String, Set<T_extract_rule>> getExtractStringRuleMap() {
  // return extractStringRuleMap;
  // }

  def getExtractPatternRuleMap: Map[T_extract_rule, Pattern] = extractPatternRuleMap

  def getExtractDoMainPatternRuleMap: Map[T_extract_rule, Pattern] = extractDoMainPatternRuleMap

  def getiPMap: Map[String, T_ip_rule] = iPMap

  def getBasicTypeRelMap: Map[Long, ListBuffer[T_basic_type_rel]] = basicTypeRelMap

  def getThemeBasicTypeRelMap: Map[Long, ListBuffer[T_theme_basic_type_rel]] = themeBasicTypeRelMap

  def getUrlRuleList: ListBuffer[T_url_rule] = urlRuleList

  def getThemeUrlRuleList: ListBuffer[T_theme_url_rule] = themeUrlRuleList

  def getExtractRuleList: ListBuffer[T_extract_rule] = extractRuleList

}
